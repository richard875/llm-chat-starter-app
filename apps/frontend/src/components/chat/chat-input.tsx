import { SendHorizonal } from "lucide-react";
import { useEffect, useRef, useState } from "react";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { useMessages } from "@/store/messages";
import { generateChatId } from "@/lib/utils";

interface ChatInputProps {
  onTypingChange: (isTyping: boolean) => void;
}

export const ChatInput = ({ onTypingChange }: ChatInputProps) => {
  const [input, setInput] = useState("");
  const {
    messages,
    isLoadingMsg,
    addMessage,
    updateLastMessage,
    currentChatId,
    setCurrentChatId,
    setIsLoadingMsg,
    refreshChatsAfterNewChat,
  } = useMessages();
  const inputRef = useRef<HTMLInputElement>(null);

  // Focus input when loading state changes to false
  useEffect(() => {
    if (!isLoadingMsg) {
      inputRef.current?.focus();
    }
    onTypingChange(isLoadingMsg);
  }, [isLoadingMsg, onTypingChange]);

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();

    if (!input.trim() || isLoadingMsg) return;

    // Generate a new chat ID if this is the first message in a new chat
    const isNewChat = !currentChatId;
    const chatId = isNewChat ? generateChatId() : currentChatId;
    setCurrentChatId(chatId);

    // Add user message
    addMessage({
      role: "user",
      content: input,
      chatId,
    });

    // Add empty assistant message that will be streamed
    addMessage({
      role: "assistant",
      content: "",
      chatId,
    });

    setInput("");
    setIsLoadingMsg(true);

    try {
      const response = await fetch("/api/chat", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          chatId,
          messages: [...messages, { role: "user", content: input }],
        }),
      });

      if (!response.ok) {
        throw new Error("Failed to get response");
      }

      const reader = response.body?.getReader();
      if (!reader) throw new Error("No reader available");

      const decoder = new TextDecoder();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split("\n");

        for (const line of lines) {
          if (line.startsWith("data: ")) {
            try {
              const data = JSON.parse(line.slice(6));
              if (data.content) {
                updateLastMessage(data.content);
              }
            } catch (e) {
              console.error("Error parsing SSE data:", e);
            }
          }
        }
      }
    } catch (error) {
      console.error("Error:", error);
      updateLastMessage("Sorry, there was an error processing your request.");
    } finally {
      setIsLoadingMsg(false);
      // Refresh chats list if this was a new chat
      if (isNewChat) refreshChatsAfterNewChat();
    }
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex w-full items-center space-x-3 rounded-full p-2 border border-primary shadow"
    >
      <Input
        ref={inputRef}
        value={input}
        onChange={(e) => setInput(e.target.value)}
        placeholder="Type your message..."
        name="message"
        className="flex-1 border-0 shadow-none !text-base focus-visible:ring-0"
        disabled={isLoadingMsg}
      />
      <Button
        type="submit"
        className="size-10 rounded-full bg-[#a3e636] border border-primary hover:bg-[#91cc33] cursor-pointer transition"
        disabled={isLoadingMsg}
      >
        <SendHorizonal className="text-primary" />
      </Button>
    </form>
  );
};
