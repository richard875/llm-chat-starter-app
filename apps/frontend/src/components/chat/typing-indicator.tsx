export const TypingIndicator = () => {
  return (
    <div className="flex items-center space-x-1 px-2 py-1">
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary [animation-delay:-0.3s]" />
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary [animation-delay:-0.15s]" />
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary" />
    </div>
  );
};
