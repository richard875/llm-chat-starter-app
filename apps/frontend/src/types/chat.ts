export interface Message {
  id?: number;
  chatId?: string;
  role: "user" | "assistant";
  content: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface Chat {
  chatId: string;
  title: string;
  createdAt?: string;
  updatedAt?: string;
}
