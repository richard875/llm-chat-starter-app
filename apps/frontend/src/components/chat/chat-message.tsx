import ReactMarkdown from "react-markdown";

import { TypingIndicator } from "@/components/chat/typing-indicator";
import { cn } from "@/lib/utils";
import type { Message } from "@/types/chat";

interface ChatMessageProps {
  message: Message;
  isTyping?: boolean;
}

export const ChatMessage = ({ message, isTyping }: ChatMessageProps) => {
  const isUser = message.role === "user";

  return (
    <div className={cn("flex mb-5", isUser ? "justify-end" : "justify-start")}>
      <div
        className={cn(
          "px-5 py-3 rounded-3xl",
          isUser
            ? "bg-primary text-primary-foreground"
            : "text-foreground border-primary border-1"
        )}
      >
        {message.content ? (
          <div className="leading-6 text-base tracking-wide">
            <ReactMarkdown>{message.content}</ReactMarkdown>
          </div>
        ) : isTyping ? (
          <TypingIndicator />
        ) : null}
      </div>
    </div>
  );
};
