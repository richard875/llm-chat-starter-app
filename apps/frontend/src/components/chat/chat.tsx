import { useEffect, useRef, useState } from "react";

import { ChatInput } from "@/components/chat/chat-input";
import { ChatMessage } from "@/components/chat/chat-message";
import { useMessages } from "@/store/messages";

export const Chat = () => {
  const { messages } = useMessages();
  const scrollAreaRef = useRef<HTMLDivElement>(null);
  const [isTyping, setIsTyping] = useState(false);

  // Scroll to bottom when messages change or typing state changes
  useEffect(() => {
    if (scrollAreaRef.current) {
      scrollAreaRef.current.scrollTop = scrollAreaRef.current.scrollHeight;
    }
  }, [messages, isTyping]);

  return (
    <div className="flex flex-col h-full">
      <div className="flex-1 overflow-y-auto" ref={scrollAreaRef}>
        {messages.length === 0 ? (
          <div className="w-full h-full pb-10 flex items-center justify-center select-none">
            <p className="text-2xl font-medium">
              Start a conversation with the AI assistant.
            </p>
          </div>
        ) : (
          <div className="max-w-screen-md mx-auto py-8 px-4">
            <div className="space-y-4">
              {messages.map((message, index) => (
                <ChatMessage
                  key={index}
                  message={message}
                  isTyping={
                    isTyping &&
                    index === messages.length - 1 &&
                    message.role === "assistant"
                  }
                />
              ))}
            </div>
          </div>
        )}
      </div>
      <div className="w-full bg-background z-20 px-5">
        <div className="max-w-screen-md mx-auto">
          <ChatInput onTypingChange={setIsTyping} />
          <p className="text-sm text-center py-2.5 text-muted-foreground select-none">
            AI can make mistakes. Please verify the information provided.
          </p>
        </div>
      </div>
    </div>
  );
};
