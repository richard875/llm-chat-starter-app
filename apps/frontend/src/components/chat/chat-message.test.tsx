import { describe, test, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import { ChatMessage } from "@/components/chat/chat-message";
import type { Message } from "@/types/chat";

// Mock react-markdown
vi.mock("react-markdown", () => ({
  default: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="markdown">{children}</div>
  ),
}));

describe("ChatMessage", () => {
  const userMessage: Message = {
    id: 1,
    role: "user",
    content: "Hello, how are you?",
    chatId: "chat-1",
    createdAt: "2023-01-01T00:00:00.000Z",
  };

  const assistantMessage: Message = {
    id: 2,
    role: "assistant",
    content: "I am doing well, thank you for asking!",
    chatId: "chat-1",
    createdAt: "2023-01-01T00:00:00.000Z",
  };

  test("renders user message correctly", () => {
    const { container } = render(<ChatMessage message={userMessage} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders assistant message correctly", () => {
    const { container } = render(<ChatMessage message={assistantMessage} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders assistant message with typing indicator", () => {
    const { container } = render(
      <ChatMessage message={assistantMessage} isTyping={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders message with empty content", () => {
    const emptyMessage: Message = {
      ...userMessage,
      content: "",
    };
    const { container } = render(<ChatMessage message={emptyMessage} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
