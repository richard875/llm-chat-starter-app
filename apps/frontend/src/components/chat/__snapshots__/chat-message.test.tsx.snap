// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatMessage > renders assistant message correctly 1`] = `
<div
  class="flex mb-7 last:mb-0 justify-start"
>
  <div
    class="min-w-0 max-w-full text-foreground"
  >
    <div
      class="leading-6.25 text-[15px] tracking-wide min-w-0 w-full"
    >
      <div
        data-testid="markdown"
      >
        I am doing well, thank you for asking!
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatMessage > renders assistant message with typing indicator 1`] = `
<div
  class="flex mb-7 last:mb-0 justify-start"
>
  <div
    class="min-w-0 max-w-full text-foreground"
  >
    <div
      class="leading-6.25 text-[15px] tracking-wide min-w-0 w-full"
    >
      <div
        data-testid="markdown"
      >
        I am doing well, thank you for asking!
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatMessage > renders message with empty content 1`] = `
<div
  class="flex mb-7 last:mb-0 justify-end"
>
  <div
    class="min-w-0 max-w-full bg-primary text-primary-foreground px-4 py-2 rounded-[20px]"
  />
</div>
`;

exports[`ChatMessage > renders user message correctly 1`] = `
<div
  class="flex mb-7 last:mb-0 justify-end"
>
  <div
    class="min-w-0 max-w-full bg-primary text-primary-foreground px-4 py-2 rounded-[20px]"
  >
    <div
      class="leading-6.25 text-[15px] tracking-wide min-w-0 w-full"
    >
      <div
        data-testid="markdown"
      >
        Hello, how are you?
      </div>
    </div>
  </div>
</div>
`;
